### PR TITLE
MODFIN-94: Add series field while creating a FiscalYear

### PR DESCRIPTION
**Pull request by javalkar-varun**
## Purpose

https://issues.folio.org/browse/MODFIN-94

## Approach

- During POST and PUT, always calculate `series`
- Add unit tests

Extra: 
- Fix `ArrayIndexOutOfBound` console error 

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/FiscalYearsApi.java [hunk 1]
```diff
@@ -22,13 +22,20 @@
...
   @Validate
   @Override
-  public void postFinanceFiscalYears(String lang, FiscalYear entity, Map<String, String> headers,
+  public void postFinanceFiscalYears(String lang, FiscalYear fiscalYear, Map<String, String> headers,
       Handler<AsyncResult<Response>> handler, Context ctx) {
     FiscalYearsHelper helper = new FiscalYearsHelper(headers, ctx, lang);
-    helper.createFiscalYear(entity)
+
+    // series should always be present
+    if (isEmpty(fiscalYear.getSeries())) {
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/FiscalYearsApi.java]: It is always a system calculated field, even if it was specified in the request, it's better to calculate than to use it

Comment (javalkar-varun) [reply src/main/java/org/folio/rest/impl/FiscalYearsApi.java]: I'm calculating and setting series on line 36, is there a better way to calculate?

Comment (KVupp) [reply src/main/java/org/folio/rest/impl/FiscalYearsApi.java]: is Empty check is unnecessary , calculate the series always

Comment (javalkar-varun) [reply src/main/java/org/folio/rest/impl/FiscalYearsApi.java]: Got it. changed.

File: src/main/java/org/folio/rest/impl/FiscalYearsApi.java [hunk 2]
```diff
@@ -22,13 +22,19 @@
...
   @Override
-  public void postFinanceFiscalYears(String lang, FiscalYear entity, Map<String, String> headers,
+  public void postFinanceFiscalYears(String lang, FiscalYear fiscalYear, Map<String, String> headers,
       Handler<AsyncResult<Response>> handler, Context ctx) {
     FiscalYearsHelper helper = new FiscalYearsHelper(headers, ctx, lang);
-    helper.createFiscalYear(entity)
+    String code = fiscalYear.getCode();
+
+    // series should always be calculated
+    fiscalYear.setSeries(code.substring(0, code.length() - FISCAL_YEAR_LENGTH));
```
Comment (grabsefx) [inline src/main/java/org/folio/rest/impl/FiscalYearsApi.java]: This lines of code repeated 2 times in POST method and PUT method. Wouldn't it be better to move this part of code into separate method something like
```
private FiscalYear getFYearWithSeries(FiscalYear fiscalYear){
   String code = fiscalYearRequest.getCode();
   return fiscalYear.withSeries(code.substring(0, code.length() - FISCAL_YEAR_LENGTH));
}
```

File: src/test/java/org/folio/rest/impl/EntitiesCrudBasicsTest.java [hunk 3]
```diff
@@ -104,6 +106,26 @@
...
+
+  @Test
+  public void testUpdateRecordEmptySeriesFY() {
+    logger.info("=== Test update {} record with empty series for FY and it should calculate series ===", TestEntities.FISCAL_YEAR);
+
+    JsonObject body = TestEntities.FISCAL_YEAR.getMockObject();
+    body.putNull("series");
+
+    verifyPut(TestEntities.FISCAL_YEAR.getEndpointWithId((String) body.remove(ID)), body, "", NO_CONTENT.getStatusCode());
+  }
```
Comment (KVupp) [inline src/test/java/org/folio/rest/impl/EntitiesCrudBasicsTest.java]: Please add a check to assert that the record sent to storage had the series populated

Comment (javalkar-varun) [reply src/test/java/org/folio/rest/impl/EntitiesCrudBasicsTest.java]: Done

File: src/main/java/org/folio/rest/impl/FiscalYearsApi.java [hunk 4]
```diff
@@ -85,4 +93,8 @@ public void deleteFinanceFiscalYearsById(String id, String lang, Map<String, Str
       .exceptionally(fail -> handleErrorResponse(handler, helper, fail));
   }
 
+  private FiscalYear getFYearWithSeries(FiscalYear fiscalYear) {
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/FiscalYearsApi.java]: this method can return a void

File: src/main/java/org/folio/rest/impl/FiscalYearsApi.java [hunk 5]
```diff
@@ -85,4 +93,8 @@ public void deleteFinanceFiscalYearsById(String id, String lang, Map<String, Str
       .exceptionally(fail -> handleErrorResponse(handler, helper, fail));
   }
 
+  private void getFYearWithSeries(FiscalYear fiscalYear) {
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/impl/FiscalYearsApi.java]: Could you rename this method since it does not return fiscalYear but set series.

**Pull request merged**